types/src/contract_event.rs

```diff
@@ -385,3 +385,184 @@ impl EventWithProof {
         Ok(())
     }
 }
+
+/// The response type for `get_event_by_version_with_proof`, which contains lower
+/// and upper bound events surrounding the requested version along with proofs
+/// for each event.
+///
+/// ### Why do we need two events?
+///
+/// If we could always get the event count _at the requested event_version_, we
+/// could return only the lower bound event. With the event count we could verify
+/// that the returned event is actually the latest event at the historical ledger
+/// view just by checking that `event.sequence_number == event_count`.
+///
+/// Unfortunately, the event count is only (verifiably) accessible via the
+/// on-chain state. While we can easily acquire the event count near the chain
+/// HEAD, historical event counts (at versions below HEAD for more than the prune
+/// window) may be inaccessible after most non-archival fullnodes have pruned past
+/// that version.
+///
+/// ### Including the Upper Bound Event
+///
+/// In contrast, if we also return the upper bound event, then we can always
+/// verify the request even if the version is past the prune window and we don't
+/// know the event_count (at event_version). The upper bound event lets us prove
+/// that there is no untransmitted event that is actually closer to the requested
+/// event_version than the lower bound.
+///
+/// For example, consider the case where there are three events at versions 10,
+/// 20, and 30. A client asks for the latest event at or below version 25. If we
+/// just returned the lower bound event, then a malicious server could return
+/// the event at version 10; the client would not be able to distinguish this
+/// response from the correct response without the event count (2) at version 25.
+///
+/// If we also return the upper bound event (the event at version 30), the client
+/// can verify that the upper bound is the next event after the lower bound and
+/// that the upper bound comes after their requested version. This proves that
+/// the lower bound is actually the latest event at or below their requested
+/// version.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+#[cfg_attr(any(test, feature = "fuzzing"), derive(Arbitrary))]
+pub struct EventByVersionWithProof {
+    pub lower_bound_incl: Option<EventWithProof>,
+    pub upper_bound_excl: Option<EventWithProof>,
+}
```

Comment:
The proof will be more self-contained if it includes a `Option<AccountStateWithProof>` that shows the latest sequence number when necessary. 

Not being the case, is that because there's no generic way os parsing the sequence number? Or is it that in case the client side has more efficient ways to know the latest sequence number?

Reply:
yeah, I figured without a way to parse the event count automatically it's not really worth it.

also the NewBlockEvent use case (which is what i'm immediately using this for) doesn't need to know the height explicitly to know if we're in the latest block or not. mostly do that so i can avoid an extra `get_account_state_with_proof(diem_root(), None, None)` call